SFM: sfm-based depth, some breadcrumbs for pycolmap

pwais opened the pull request:
In this PR:
 * Generate depth map images from COLMAP 3D SfM points.  Tested with `nerfacto` and `depth-nerfacto` on poster scene.  (TODO: test others?  Can I create a jupyter notebook or something to reproduce a DS-NeRF-like few-image result?).  
 * ~Replace legacy COLMAP parsing code with pycolmap.  Requires upgrade to COLMAP 3.8 in docker.~  Punted to https://github.com/nerfstudio-project/nerfstudio/issues/1480 due to Windows issues.
 * Add documentation & tests about quaternion formats-- some datasets use scalar-first and some uses scalar-last.  This part also helps confirm that the `pycolmap` refactor should produce identical transforms to what exists prior to this change.
 * Improve `Dockerfile` ergonomics:
    * Add `CUDA_ARCHITECTURES` arg-- I can't actually even build arches 90 and 89 locally (I get compile errors)
    * Add pycolmap from source, there will be a wheel next release (see cited ticket)
    * Pin `tiny-cuda-nn` version, I was getting some weird errors otherwise.
    * Add dockerignore to help create fast builds and clean releases.  Maybe we start a nerfstudio dockerhub account?
 * ~Closes https://github.com/nerfstudio-project/nerfstudio/issues/263 (looks like that issue might be prematurely closed)~


Here are point clouds for nerfacto on poster with and without depth.  They look pretty similar:
 * https://drive.google.com/drive/folders/1ILCkMon41Dtfln5pUHxNKm-ndm7wmhWf?usp=share_link

pwais reviewed (commented):
Questions:
 * Is there a good procedure for documenting a training run with artifacts?  Right now I'm just using the poster scene and docker image locally... I don't think
 * I was thinking of adding a unit test of the COLMAP depth maps, in particular that includes depth-colored RGB images.  I checked this manually but an automated check with viz would probably be very nice to have.  Can I use the lego scene for this perhaps?  It will require adding a couple more static assets / test images to the repo & build process.

tancik commented:
Question, is pycolmap supported on windows (without using WSL)? It seems to be suspiciously missing from [this list](https://github.com/colmap/pycolmap#:~:text=Wheels%20for%20Linux%20(Python%203.6/7/8/9)%20and%20macOS%2010%20%26%2011%20(Python%203.7/8/9)%20can%20be%20install%20using%20pip%3A)
If so, I would be worried about transitioning to it. There is a sizable fraction of less technical nerfstudio users on windows that don't use WSL.

tancik commented on nerfstudio/process_data/polycam_utils.py:
```diff
@@ -215,6 +215,7 @@ def process_depth_maps(
         summary_log.append(f"Started with {num_processed_depth_maps} images")
 
     # Downscale depth maps
+    # TODO(reviewer): should this be nearest neighbor???
```
It should be, good catch.

pwais commented:
derrrrp

https://github.com/colmap/pycolmap/blob/master/.github/workflows/build-new.yml#L13

well mebbe we can make a windows wheel real easy peezy.  github does windows after all ...

pwais commented:
cc https://github.com/colmap/pycolmap/issues/34

pwais commented:
cc https://github.com/pwais/pycolmap/pull/1

pwais commented:
@tancik FWIW so I've been trying to build a windows pycolmap wheel... With modern Core i7 & good internet, it's taking well over two hours (and still going) to build & install via `vcpkg`, which appears to be their recommended install when not using the GUI. (see https://colmap.github.io/install.html#id3  and note here: https://github.com/colmap/pycolmap/issues/34#issuecomment-1437182649 ).

Do you happen to know how any of the windows users are using Nerfstudio + COLMAP?  and how the install COLMAP?  

I'll keep poking.. if windows users are using conda, then perhaps there's a way to bridge vcpkg and conda https://github.com/conda-forge/vcpkg-feedstock 

I take it that telling windows users to use a docker image (and set up a nerfstudio dockerhub account w/ standard image) is a bad idea?

tancik commented:
> Do you happen to know how any of the windows users are using Nerfstudio + COLMAP? and how the install COLMAP?

Personally no, but there are many windows users on the discord.

> I take it that telling windows users to use a docker image (and set up a nerfstudio dockerhub account w/ standard image) is a bad idea?

Yea I'd prefer not to take that approach.

While the colmap stuff is being figured out, do you think it is worth merging in some of the other changes (.ignore files, NN depth downsampling)? Also I noticed you added a `create_sfm_depth` in a previous commit, can we use that instead of pycolmap for now?

pwais commented:
ok i'll give pycolmap windows wheels one more push and then if it's gonna be too long i'll factor the pycolmap stuff out and ticket it

pwais commented:
Ok I removed the `pycolmap` requirement and added depth debug images, which look correct for the poster scene:
![frame_00110 png debug](https://user-images.githubusercontent.com/218145/220842962-34685b5c-b2ea-4e96-8932-5a8b903aa220.jpg)
![frame_00205 png debug](https://user-images.githubusercontent.com/218145/220842909-285d9403-e6de-48c4-b952-16830555fb9f.jpg)
![frame_00176 png debug](https://user-images.githubusercontent.com/218145/220842921-f95f4a76-38f3-497f-a63e-9c853523e9aa.jpg)

in `test/data`  could we add a perhaps downsized version of the lego synthetic dataset as a test fixture?  wouldn't it be a helpful test to have `ns-process-data images` run on a fixture like that? Or at least I was thinking of adding a unit test for the depth image conversion.

pwais commented on nerfstudio/process_data/polycam_utils.py:
```diff
@@ -216,7 +216,9 @@ def process_depth_maps(
     summary_log.append(
-        process_data_utils.downscale_images(depth_dir, num_downscales, folder_name="depth", verbose=verbose)
+        process_data_utils.downscale_images(
+            depth_dir, num_downscales, folder_name="depths", nearest_neighbor=True, verbose=verbose
```
NB: this is not only fixed to use nearest-neighbor, but I believe the `folder_name` was actually broken too?  it appears that `depth` is the 1x directory, and then `depths_{s}` is the dir for downscale by `s`

pwais commented:
@tancik Ok I've pulled `main` and `ns-dev-test` passes in the Docker container.  Does this look good enough to merge?

tancik commented:
@Zunhammer Do you mind reviewing the docker changes in this PR?

tancik commented on nerfstudio/data/utils/colmap_utils_3p.py:
```diff
@@ -1,3 +1,11 @@
+"""
+This file copied with small modifications from:
```
Why rename the file?

pwais commented:
renamed because there is already another colmap_utils and its less
confusing this way no?

On Fri, Feb 24, 2023 at 14:18 Matthew Tancik ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In nerfstudio/data/utils/colmap_utils_3p.py
> <https://github.com/nerfstudio-project/nerfstudio/pull/1457#discussion_r1117738304>
> :
>
> > @@ -1,3 +1,11 @@
> +"""
> +This file copied with small modifications from:
>
> Why rename the file?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nerfstudio-project/nerfstudio/pull/1457#pullrequestreview-1314326323>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABVIIKWJU5UPMAP25DMLADWZEXTFANCNFSM6AAAAAAVBH44UQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

tancik commented:
> renamed because there is already another colmap_utils and its less confusing this way no?
> […](#)
> On Fri, Feb 24, 2023 at 14:18 Matthew Tancik ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In nerfstudio/data/utils/colmap_utils_3p.py <[#1457 (comment)](https://github.com/nerfstudio-project/nerfstudio/pull/1457#discussion_r1117738304)> : > @@ -1,3 +1,11 @@ +""" +This file copied with small modifications from: Why rename the file? — Reply to this email directly, view it on GitHub <[#1457 (review)](https://github.com/nerfstudio-project/nerfstudio/pull/1457#pullrequestreview-1314326323)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABVIIKWJU5UPMAP25DMLADWZEXTFANCNFSM6AAAAAAVBH44UQ> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Fair enough. Maybe a more specific name would be good then, ie `colmap_parsing_utils.py`. I'm not sure with `p3` means.

Zunhammer commented:
> @Zunhammer Do you mind reviewing the docker changes in this PR?

@tancik Sure, looks fine to me.

In detail: Personally I would not add the verbose flag to the pip installs as it's already verified before I do any changes to the Dockerfile but doesn't hurt. Pinning version of tinycudann makes sense, limiting number of threads to number of cores for building with make also.

pwais commented:
> Fair enough. Maybe a more specific name would be good then, ie `colmap_parsing_utils.py`. I'm not sure with `p3` means.

3p as in "third party" since it's third party code.  ok will rename.  

@Zunhammer ok ill remove the verbose sure.  might consider doing `pip -e .[all]` on nerfstudio to include the `[dev]` and `[docs]` stuff?  but mebbe that's for a future change.

pwais commented:
hmm I had thought we could add LoFTR thru hloc to get denser depth, but it looks like they dont have that in their pipeline code yet.  nevertheless, it's something we could add for nerfstudio on top of this change.  should be useful to have for benchmarking

![Screen Shot 2023-02-25 at 3 08 29 PM](https://user-images.githubusercontent.com/218145/221383813-2090a1e8-0a5e-4af8-8cc2-8c2af7d7592d.png)
![Screen Shot 2023-02-25 at 3 12 12 PM](https://user-images.githubusercontent.com/218145/221383815-b61e3622-2c8e-4c8c-956f-7a02cc5a3f36.png)

tancik reviewed (approved):
LGTM, thanks for adding this functionality.

jkulhanek commented:
This PR introduced some bugs. The `ns-process-data images` stopped working and --colmap-model-path was ignored. I fixed it here: a075b36db9c88186361a27cd1cc9f269d5f7d3cf.

Also, the new code is missing type annotations. The annotations for COLMAP-related functions are missing altogether and types for COLMAP's Cameras were removed. What is the benefit of using "pycolmap" over the previous code? To me the previous code was perfectly valid and I would prefer it over adding a new dependency.

pwais commented:
> This PR introduced some bugs. The `ns-process-data images` stopped working and --colmap-model-path was ignored. I fixed it here: [a075b36](https://github.com/nerfstudio-project/nerfstudio/commit/a075b36db9c88186361a27cd1cc9f269d5f7d3cf).

whoops sorry!  not sure how that got pasted in there.  yeah this module could really use some test coverage.  i'm a total n00b so i largely rely on unit tests / `ns-dev-test` to tell me if something is good.


> Also, the new code is missing type annotations. 
`colmap_utils.py` now `colmap_parsing_utils.py` never had type annotations as far as i can tell.  

> The annotations for COLMAP-related functions are missing altogether and types for COLMAP's Cameras were removed. What is the benefit of using "pycolmap" over the previous code? To me the previous code was perfectly valid and I would prefer it over adding a new dependency.

In previous main, there were *two* copies of third party colmap code.  That was confusing, and add to that both files had names `colmap_utils.py` -- very very confusing.  The one that this change retained had no major modifications when i tried to compare it to the COLMAP source.  The other, it was unclear what had been changed, what had not, and importantly what the license and attribution was.  

Why `pycolmap`?
 * I primarily added it for the 2D and 3D point utils.  COLMAP has a somewhat complicated and poorly-interfaced internal database; pycolmap provides a cleaner, peer-reviewed interface to that data.  
 * Need to be very careful about the handling of rotation matrices, especially because nerfstudio uses the OpenGL frame convention.  I could trust pycolmap did this correctly.  There's a new test added to bring attention to the issue: https://github.com/nerfstudio-project/nerfstudio/blob/main/tests/process_data/test_misc.py
 * There had been requests for it in the github issues

But also, pycolmap helps avoid the case we had here where people copy-paste and hack up COLMAP parsing code which leads to confusion, potential bugs, and potential licensing issues.  Personally, I think it's fine to copy-paste & hack things up for research code / alpha-quality code, but there needs to be some clean-up or unit testing for community-supported software else everybody can get confused.

jkulhanek commented:
Some more bugs were fixed here #1527

pwais commented:
doh! yeah that's a pretty typical consequence of having copy-pasta (e.g. between process images and videos) and no automated testing.  i dont know what the repo policy is on git LFS but would be good to add a test of e.g. sfm and training on a small nerf lego scene end-to-end.  (the repo itself seems like git history might be screwed up-- .git is like 70MB yet there are very few binary assets in the checkout).   or even if `ns-dev-test` does an end-to-end test on dev machine that has GPU, since CI budget is not big

silver-obelisk commented:
> 咚！是的，这是复制粘贴（例如在处理图像和视频之间）并且没有自动测试的一个非常典型的结果。我不知道 Git LFS 上的回购策略是什么，但最好添加例如 SFM 的测试和端到端小型 nerf 乐高场景的训练。（存储库本身似乎 git 历史记录可能被搞砸了-- .git 就像 70MB，但结帐中的二进制资产很少）。或者即使在具有GPU的开发计算机上进行端到端测试，因为CI预算不大`ns-dev-test`

Hi,I found you add sfm depth to nerfstudio, but  I donnot understand how to use it. Is it only can used in depth-nerfacto ? or any other method such as instant ngp? how does it work, just like ds-nerf to use sparse point depth and set other pixel depth=0? Which depth loss you used?  If you can tell me more detail that will be so useful for me. Thank you so much!
